src/main/kotlin/de/itemis/mps/gradle/runmigrations/Plugin.kt

```diff
@@ -0,0 +1,83 @@
+package de.itemis.mps.gradle.runmigrations
+
+import de.itemis.mps.gradle.*
+import net.swiftzer.semver.SemVer
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Copy
+import org.gradle.kotlin.dsl.withGroovyBuilder
+import java.io.File
+import javax.inject.Inject
+
+open class MigrationExecutorPluginExtensions @Inject constructor(of: ObjectFactory, project: Project) : BasePluginExtensions(of, project) {
+    val force: Property<Boolean> = of.property(Boolean::class.java).convention(false)
+}
+@Suppress("unused")
+open class RunMigrationsMpsProjectPlugin : Plugin<Project> {
+    override fun apply(project: Project) {
+        project.run {
+            val extension = extensions.create("runMigrations", MigrationExecutorPluginExtensions::class.java)
+            afterEvaluate {
+                val mpsLocation = extension.mpsLocation.convention{ File(project.buildDir, "mps") }.map { it.asFile }.get()
+                val projectLocation = extension.projectLocation.getOrElse { throw GradleException("No project path set") }
+                if(!file(projectLocation).exists()) {
+                    throw GradleException("The path to the project doesn't exist:$projectLocation")
+                }
+                val forceMigration = extension.force
+                
+                val mpsVersion = extension.getMPSVersion()
+                val parsedMPSVersion = SemVer.parse(mpsVersion)
+                if(forceMigration.getOrElse(false) && (parsedMPSVersion.major < 2021 || (parsedMPSVersion.major == 2021 && parsedMPSVersion.minor < 3))) {
+                    throw GradleException("The force migration flag is only supported for 2021.3.0 and higher.")
+                }
+                
+                val resolveMps: Task = if (extension.mpsConfig.isPresent) {
+                    tasks.create("resolveMpsForMigrations", Copy::class.java) {
+                        from({extension.mpsConfig.get().resolve().map { zipTree(it) }})
+                        into(mpsLocation)
+                    }
+                } else {
+                    tasks.create("resolveMpsForMigrations")
+                }
+                
+                tasks.register("runMigrations") {
```

Comment:
I found it's useful to register tasks right when the plugin is applied, before `project.afterEvaluate`, and then configure them in `project.afterEvaluate`.

However, in general, it's best to do as little work as possible in `project.afterEvaluate` because it's called late in the lifecycle. If the user wants to adjust or override something that our plugin does in `afterEvaluate`, they may have to resort to adding their own `project.afterEvaluate` callback and it gets ugly quickly.